### PR TITLE
Migrate the legacy AssetRequestHandlerUnitTests

### PR DESCRIPTION
**Pull request by junbo75**
Signed-off-by: Junbo Liang <68558268+junbo75@users.noreply.github.com>

## What does this PR do?

Migrated the legacy AssetRequestHandlerUnitTests and broke it into multiple tests

## How was this PR tested?

```
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from AssetRequestHandlerUnitTests
[ RUN      ] AssetRequestHandlerUnitTests.RequestToProcessAssetNotExistInDatabaseOrQueue_RequestSent_RequestHandled
[       OK ] AssetRequestHandlerUnitTests.RequestToProcessAssetNotExistInDatabaseOrQueue_RequestSent_RequestHandled (720 ms)
[ RUN      ] AssetRequestHandlerUnitTests.RequestToCreateCompileGroup_RequestSent_RequestHandled
[       OK ] AssetRequestHandlerUnitTests.RequestToCreateCompileGroup_RequestSent_RequestHandled (303 ms)
[ RUN      ] AssetRequestHandlerUnitTests.RequestToCompileAsset_RequestSent_RequestHandled
[       OK ] AssetRequestHandlerUnitTests.RequestToCompileAsset_RequestSent_RequestHandled (315 ms)
[ RUN      ] AssetRequestHandlerUnitTests.RequestToProcessFileOnDiskButNotInQueue_RequestSent_RequestHandled
[       OK ] AssetRequestHandlerUnitTests.RequestToProcessFileOnDiskButNotInQueue_RequestSent_RequestHandled (318 ms)
[ RUN      ] AssetRequestHandlerUnitTests.TestMultipleInFlightRequests_RequestsSent_RequestsHandled
[       OK ] AssetRequestHandlerUnitTests.TestMultipleInFlightRequests_RequestsSent_RequestsHandled (315 ms)
[----------] 5 tests from AssetRequestHandlerUnitTests (1972 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (1973 ms total)
[  PASSED  ] 5 tests.
```

### CODE REVIEW

File: Code/Tools/AssetProcessor/native/unittests/AssetRequestHandlerUnitTests.h [hunk 1]
```diff
@@ -6,18 +6,34 @@
...
-        AssetRequestHandlerUnitTests();
-        virtual void StartTest() override;
+
+    protected:
+        void SetUp() override;
+        void TearDown() override;
+
+        AZStd::unique_ptr<MockAssetRequestHandler> m_requestHandler;
+
+        AssetProcessor::MockConnectionHandler m_connection;
```
Comment (jackalbe) [inline Code/Tools/AssetProcessor/native/unittests/AssetRequestHandlerUnitTests.h:32]: Should MockConnectionHandler m_connection be a AZStd::unique_ptr like m_requestHandler?

**Review by jackalbe — APPROVED** [review]

**Review by AMZN-stankowi — APPROVED** [review]

File: Code/Tools/AssetProcessor/native/unittests/AssetRequestHandlerUnitTests.cpp [hunk 2]
```diff
@@ -73,11 +74,21 @@ namespace AssetProcessor
...
 }
 
+AssetRequestHandlerUnitTests::AssetRequestHandlerUnitTests()
+    : UnitTest::AssetProcessorUnitTestBase()
+{
+}
+
+AssetRequestHandlerUnitTests::~AssetRequestHandlerUnitTests()
+{
+}
```
Comment (amzn-mike) [inline Code/Tools/AssetProcessor/native/unittests/AssetRequestHandlerUnitTests.cpp:84]: Why were these needed? Especially the destructor

**Review by amzn-mike — APPROVED** [review]

File: Code/Tools/AssetProcessor/native/unittests/AssetRequestHandlerUnitTests.cpp (hunk 2 above)
Comment (junbo75) [reply Code/Tools/AssetProcessor/native/unittests/AssetRequestHandlerUnitTests.cpp]: The destructor is required because of the forward declaration and unique pointer for MockAssetRequestHandler and MockConnectionHandler. Otherwise the compiler generates a default one, and it needs a complete declaration of these two classes, which caused the Linux build failure: https://jenkins.build.o3de.org/blue/organizations/jenkins/O3DE/detail/PR-12653/2/pipeline
